proposals/0000-integrations.md

```diff
@@ -0,0 +1,171 @@
+- Start Date: 03-07-2022
+- Reference Issues: <!-- related issues, otherwise leave empty -->
+- Implementation PR: [(git branch, runnable but not yet a PR)](https://github.com/withastro/astro/compare/wip-integrations-4?expand=1)
+
+# Summary
+
+An integration system for extending Astro.
+
+# Example
+
+```js
+// astro.config.js
+export default ({
+	integrations: [
+		import('@astrojs/vue'),
+		import('@astrojs/tailwind'),
+		[import('@astrojs/partytown'), {/* options */}],
+	],
+});
+```
+
+```js
+// Static imports are also supported, for type checking.
+import vuePlugin from '@astrojs/vue';
+import tailwindPlugin from '@astrojs/tailwind';
+import partytownPlugin from '@astrojs/partytown';
+
+export default ({
+	integrations: [
+		vuePlugin(),
+		tailwindPlugin(),
+		partytownPlugin({/* options */})
+	],
+});
```

Comment:
This is just personal preference, but I much prefer this syntax as it feels familiar to importing components and styles in an Astro component

Is there a benefit to also supporting the `import("@astrojs/vue")` syntax as well?  I've always been a bit thrown off with tools that have multiple ways for config (different file types, extensions, CJS vs. ESM, etc), I'd love to see one supported approach if that's feasible

Reply:
To be honest... I only see cons with that `import` syntax 😆 
1. Allowing the `[nestedArraySyntax, { /** with options */ }]` removes type safety while configuring. You'd need to import a typecast for that `{ optionsObject }` to offer any intellisense.
2. Allowing users to `import` without an `await` teaches a pretty dangerous pattern. As a user that hasn't seen dynamic imports, I'd assume I could just `import` without `await` anywhere in my Astro templates too. This is obviously a pretty dangerous assumption!
3. Allowing 2 different syntaxes could introduce all sorts of confusion for tutorial writers and boilerplate users.

I say we use the static `import` at the top of the file and ditch the second syntax entirely.

Reply:
lol, I kind of knew this would be unpopular even as I wrote it :) This isn't a hill I'll die on, but to at least make the case for it:

Static imports are more verbose, requiring code in two different places (the import and the `integrations` array entry). If I imagine someone running through a tutorial, I find either of the first two examples much easier to follow:

```diff
integrations: [
+  "tailwind"
]
```

```diff
integrations: [
+  import("@astrojs/tailwind")
]
```


```diff
// At the top of your file:
+ import tailwindPlugin from '@astrojs/tailwind';

// In the config:
integrations: [
+  tailwindPlugin(),
]
```

Yes, I know, the irony isn't lost on me that this is the same argument for inline asset references in the template, vs. ESM imports of assets 🙈 . 

I don't think it's impossible to support two different styles here, but I agree its extra complexity that we shouldn't take on lightly or without good reason. If tutorial authors felt like they had to document both APIs, then that would be a good reason not to pursue this.

Also worth calling out that an awesome `astro add tailwind` CLI command that code-mods your config file for you would solve the "user following a tutorial" problem, which is the main problem I was trying to design for here.